File: layout.py
"""Auto-layout: pack components tightly into a region/row grid.

Input:
  - `Diagram` with components/regions/edges (positions ignored)
  - `LAYOUT`  — dict of region_id → list[row], where row is a list of cell-ids
  - `EXTERNAL` — components placed relative to a target (e.g. user above jupyter)

Output (in place):
  - Each component's `pos` is set to its (cx, cy) center
  - Each region's `bounds` is set tight around its contents
  - Each edge with `route != "auto"` and empty `via` gets routing waypoints
  - Diagram width/height set to enclose everything

Whitespace minimisation principles:
  1. Per-row height = max(component height + label height) across all regions
     so rows align horizontally across regions (cross-region edges run flat).
  2. Per-region width = max(row width) — region only as wide as it needs.
  3. Component cell width = max(icon width, longest label width) + padding.
  4. Same gap between rows / between cells / between regions, but each is
     a tunable parameter so we don't over- or under-pad anywhere.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

from model import Component, Diagram, Edge, LABEL_HEIGHT


# ── Sizing ─────────────────────────────────────────────────────────────
# Empirically-tuned character widths (avg over Latin + Vietnamese diacritics
# at the given pixel size). Off by < 5% which is plenty for layout.
_CHAR_W_NAME = 7.6        # 14px component name
_CHAR_W_SUB  = 6.4        # 11.5px subtitle
_LABEL_GAP   = 6
_LINE_HEIGHT = 18

_ICON_DIMS: dict[str, tuple[int, int]] = {
    "circle":     (76, 76),
    "cube":       (80, 80),
    "cube-big":   (100, 100),   # body; halo adds ~16px around but we don't count it for spacing
    "box":        (70, 70),
    "cylinder":   (70, 70),
    "hex":        (70, 64),     # flat-top hexagon
    "annotation": (40, 32),
}


@dataclass
class Cell:
    w: int
    h: int


def _text_w(s: str, char_w: float) -> int:
    return int(len(s) * char_w)


def cell_size(c: Component, h_pad: int = 8, v_pad: int = 10) -> Cell:
    iw, ih = _ICON_DIMS[c.shape]
    name_w = _text_w(c.name,     _CHAR_W_NAME)
    sub_w  = _text_w(c.subtitle, _CHAR_W_SUB)
    label_w = max(name_w, sub_w)
    label_h = _LABEL_GAP + _LINE_HEIGHT * 2
    return Cell(
        w=max(iw, label_w) + h_pad * 2,
        h=ih + label_h + v_pad,
    )


# ── Layout ─────────────────────────────────────────────────────────────
def layout(
    diagram: Diagram,
    region_layout: dict[str, list[list[str]]],
    external: dict[str, dict] | None = None,
    *,
    region_gap: int = 36,
    row_gap: int = 28,
    cell_gap: int = 18,
    region_padding_x: int = 18,
    region_padding_y: int = 22,
    canvas_margin: int = 18,
) -> None:
    sizes = {c.id: cell_size(c) for c in diagram.components}
    region_ids = list(region_layout.keys())

    # Per-row height (across all regions) → consistent Y for cross-region rows.
    max_rows = max(len(rs) for rs in region_layout.values())
    row_heights: list[int] = []
    for i in range(max_rows):
        h = 0
        for rid in region_ids:
            rows = region_layout[rid]
            if i < len(rows):
                for cid in rows[i]:
                    if cid in sizes:
                        h = max(h, sizes[cid].h)
        row_heights.append(h or 100)

    # Per-region width = max(row width).
    region_widths: dict[str, int] = {}
    for rid, rows in region_layout.items():
        w = 0
        for row in rows:
            row_w = sum(sizes[cid].w for cid in row) + (len(row) - 1) * cell_gap
            w = max(w, row_w)
        region_widths[rid] = w + region_padding_x * 2

    # External components reserve vertical space above the first row.
    external = external or {}
    ext_above_height = 0
    for eid, spec in external.items():
        if spec.get("above") and eid in sizes:
            ext_above_height = max(ext_above_height, sizes[eid].h + spec.get("gap", 60))

    # Row centre-Y, cumulative.
    row_ys: list[int] = []
    y = canvas_margin + ext_above_height
    for h in row_heights:
        y += h // 2
        row_ys.append(y)
        y += h // 2 + row_gap
    last_y_after_rows = y - row_gap + region_padding_y // 2  # for canvas height

    # Region centre-X, cumulative.
    region_xs: dict[str, int] = {}
    x = canvas_margin
    for rid in region_ids:
        w = region_widths[rid]
        x += w // 2
        region_xs[rid] = x
        x += w // 2 + region_gap
    canvas_right = x - region_gap + canvas_margin

    # Place each component at its grid cell.
    for rid, rows in region_layout.items():
        rx = region_xs[rid]
        for i, row in enumerate(rows):
            ry = row_ys[i]
            total_w = sum(sizes[cid].w for cid in row) + (len(row) - 1) * cell_gap
            cursor = rx - total_w // 2
            for cid in row:
                cw = sizes[cid].w
                diagram.get(cid).pos = (cursor + cw // 2, ry)
                cursor += cw + cell_gap

    # Place external (above a target).
    for eid, spec in external.items():
        if "above" in spec:
            target = diagram.get(spec["above"])
            gap = spec.get("gap", 60)
            target_top = target.pos[1] - sizes[spec["above"]].h // 2
            diagram.get(eid).pos = (target.pos[0],
                                    target_top - gap - sizes[eid].h // 2)

    # Region bounds — hug the cells with region_padding.
    for rid, rows in region_layout.items():
        rx = region_xs[rid]
        rw = region_widths[rid]
        n = len(rows)
        top = row_ys[0] - row_heights[0] // 2 - region_padding_y
        bot = row_ys[n - 1] + row_heights[n - 1] // 2 + region_padding_y
        for r in diagram.regions:
            if r.id == rid:
                r.bounds = (rx - rw // 2, top, rw, bot - top)
                break

    # Auto-route edges that asked for it (route != "auto" or empty via).
    _route_edges(diagram, row_ys, row_heights)

    # Final canvas — fit content + margin.
    diagram.width  = canvas_right
    diagram.height = max(c.pos[1] + sizes[c.id].h // 2 for c in diagram.components) + canvas_margin


# ── Edge routing helpers ───────────────────────────────────────────────
def _route_edges(diagram: Diagram, row_ys: list[int], row_heights: list[int]) -> None:
    for e in diagram.edges:
        if e.via:
            continue  # respect explicit waypoints
        if e.route == "over":
            e.via = _route_over(diagram, e, row_ys, row_heights)
        elif e.route == "under":
            e.via = _route_under(diagram, e, row_ys, row_heights)
        # "auto" produces no via; the renderer's auto-elbow handles it.


def _route_over(diagram: Diagram, e: Edge, row_ys: list[int],
                row_heights: list[int]) -> list[tuple[int, int]]:
    """Snap the across-segment to the middle of the actual content gap above
    the source row — accounting for label area below row above's icons."""
    src = diagram.get(e.src)
    dst = diagram.get(e.dst)
    sp = src.anchor(e.src_anchor)
    dp = dst.anchor(e.dst_anchor)

    # Locate source row.
    src_y = src.pos[1]
    src_row = min(range(len(row_ys)), key=lambda i: abs(row_ys[i] - src_y))

    if src_row == 0:
        # No row above — route above everything
        top_of_content = min(c.pos[1] - c.half[1] for c in diagram.components
                             if c.shape != "annotation")
        over_y = max(14, top_of_content - 30)
        return [(sp[0], over_y), (dp[0], over_y)]

    TOL = 5
    prev_row_y = row_ys[src_row - 1]
    curr_row_y = row_ys[src_row]
    prev_cells = [c for c in diagram.components if abs(c.pos[1] - prev_row_y) < TOL]
    curr_cells = [c for c in diagram.components if abs(c.pos[1] - curr_row_y) < TOL]

    prev_bottom = max((c.pos[1] + c.half[1] + LABEL_HEIGHT.get(c.shape, 0)
                       for c in prev_cells), default=prev_row_y)
    curr_top    = min((c.pos[1] - c.half[1] for c in curr_cells), default=curr_row_y)
    over_y = (prev_bottom + curr_top) // 2

    return [(sp[0], over_y), (dp[0], over_y)]


def _route_under(diagram: Diagram, e: Edge, row_ys: list[int],
                 row_heights: list[int]) -> list[tuple[int, int]]:
    """Snap the across-segment to the middle of the actual content gap below
    the source row — accounting for label area which extends below the icon."""
    src = diagram.get(e.src)
    dst = diagram.get(e.dst)
    sp = src.anchor(e.src_anchor)
    dp = dst.anchor(e.dst_anchor)

    # Locate source row.
    src_y = src.pos[1]
    src_row = min(range(len(row_ys)), key=lambda i: abs(row_ys[i] - src_y))

    if src_row + 1 >= len(row_ys):
        return [(sp[0], max(sp[1], dp[1]) + 36), (dp[0], max(sp[1], dp[1]) + 36)]

    TOL = 5
    src_row_y = row_ys[src_row]
    next_row_y = row_ys[src_row + 1]
    src_cells  = [c for c in diagram.components if abs(c.pos[1] - src_row_y)  < TOL]
    next_cells = [c for c in diagram.components if abs(c.pos[1] - next_row_y) < TOL]

    row_bottom = max((c.pos[1] + c.half[1] + LABEL_HEIGHT.get(c.shape, 0)
                      for c in src_cells), default=src_row_y)
    next_top   = min((c.pos[1] - c.half[1] for c in next_cells), default=next_row_y)
    under_y = (row_bottom + next_top) // 2

    return [(sp[0], under_y), (dp[0], under_y)]

File: packages/js/src/layout.ts
/**
 * Auto-layout — JS port of `packages/python/src/kymo/layout.py`.
 *
 * Two layout strategies, both writing component `pos` (and region `bounds`)
 * in place:
 *   - `layout()` — grid mode: pack components into a region/row grid so rows
 *     line up across regions (cross-region same-row edges run flat).
 *   - `applyLayoutTree()` — Figma-style auto-layout: each frame hugs its
 *     contents, children flow along the frame axis with a fixed gap, cross
 *     axis centre-aligned. `minimizeCrossings()` barycenter-reorders children
 *     first so siblings land in a crossing-minimised order.
 *
 * This is an independent TypeScript implementation kept at parity with the
 * Python algorithm — not a transpile. Output need not be byte-identical to
 * Python's (the two renderers differ); it must produce equivalent geometry.
 */
import {
  LABEL_HEIGHT, SHAPE_HALF, componentHalf, anchor, resolveAnchors,
  getComponent, type Component, type Diagram, type Edge, type Point,
} from "./model.js";

// ── Layout tree node (mirrors Python's tuple shapes) ───────────────────
//   { t: "id", id }
//   { t: "group", dir, children }
//   { t: "group", dir, children, padding }   ← 4-tuple variant (cluster pad)
export type LayoutNode =
  | { t: "id"; id: string }
  | { t: "group"; dir: "horizontal" | "vertical"; children: LayoutNode[]; padding?: Point };

export function idNode(id: string): LayoutNode {
  return { t: "id", id };
}
export function groupNode(
  dir: "horizontal" | "vertical", children: LayoutNode[], padding?: Point,
): LayoutNode {
  return padding ? { t: "group", dir, children, padding } : { t: "group", dir, children };
}

export interface RegionLayout {
  [regionId: string]: string[][];
}
export interface ExternalSpec {
  [componentId: string]: { above?: string; gap?: number };
}

// ── Sizing ─────────────────────────────────────────────────────────────
const CHAR_W_NAME = 7.6;   // 14px component name
const CHAR_W_SUB = 6.4;    // 11.5px subtitle
const LABEL_GAP = 6;
const LINE_HEIGHT = 18;

// Derive icon (w, h) from SHAPE_HALF so new shapes don't drift; annotation
// keeps a (40, 32) label budget though SHAPE_HALF treats it as (0, 0).
const ICON_DIMS: Record<string, Point> = (() => {
  const out: Record<string, Point> = {};
  for (const [shape, [hw, hh]] of Object.entries(SHAPE_HALF)) out[shape] = [2 * hw, 2 * hh];
  out["annotation"] = [40, 32];
  return out;
})();

function iconDims(shape: string): Point {
  return ICON_DIMS[shape] ?? [0, 0];
}

function textW(s: string, charW: number): number {
  return Math.trunc(s.length * charW);
}

export interface Cell {
  w: number;
  h: number;
}

export function cellSize(c: Component, hPad = 8, vPad = 10): Cell {
  const [iw, ih] = iconDims(c.shape);
  const nameW = textW(c.name, CHAR_W_NAME);
  const subW = textW(c.subtitle, CHAR_W_SUB);
  const labelW = Math.max(nameW, subW);
  const labelH = LABEL_GAP + LINE_HEIGHT * 2;
  return { w: Math.max(iw, labelW) + hPad * 2, h: ih + labelH + vPad };
}

// ── Grid layout ──────────────────────────────────────────────────────────
export interface LayoutOptions {
  regionGap?: number;
  rowGap?: number;
  cellGap?: number;
  regionPaddingX?: number;
  regionPaddingY?: number;
  canvasMargin?: number;
}

export function layout(
  diagram: Diagram,
  regionLayout: RegionLayout,
  external: ExternalSpec | null = null,
  opts: LayoutOptions = {},
): void {
  const regionGap = opts.regionGap ?? 36;
  const rowGap = opts.rowGap ?? 28;
  const cellGap = opts.cellGap ?? 18;
  const regionPaddingX = opts.regionPaddingX ?? 18;
  const regionPaddingY = opts.regionPaddingY ?? 22;
  const canvasMargin = opts.canvasMargin ?? 18;

  const sizes = new Map<string, Cell>(diagram.components.map((c) => [c.id, cellSize(c)]));
  const regionIds = Object.keys(regionLayout);

  // Per-row height across all regions → consistent Y for cross-region rows.
  const maxRows = Math.max(...regionIds.map((rid) => regionLayout[rid].length));
  const rowHeights: number[] = [];
  for (let i = 0; i < maxRows; i++) {
    let h = 0;
    for (const rid of regionIds) {
      const rows = regionLayout[rid];
      if (i < rows.length) {
        for (const cid of rows[i]) {
          const s = sizes.get(cid);
          if (s) h = Math.max(h, s.h);
        }
      }
    }
    rowHeights.push(h || 100);
  }

  // Per-region width = max(row width).
  const regionWidths = new Map<string, number>();
  for (const rid of regionIds) {
    let w = 0;
    for (const row of regionLayout[rid]) {
      const rowW = row.reduce((acc, cid) => acc + (sizes.get(cid)?.w ?? 0), 0)
        + (row.length - 1) * cellGap;
      w = Math.max(w, rowW);
    }
    regionWidths.set(rid, w + regionPaddingX * 2);
  }

  // External components reserve vertical space above the first row.
  const ext = external ?? {};
  let extAboveHeight = 0;
  for (const [eid, spec] of Object.entries(ext)) {
    if (spec.above && sizes.has(eid)) {
      extAboveHeight = Math.max(extAboveHeight, sizes.get(eid)!.h + (spec.gap ?? 60));
    }
  }

  // Row centre-Y, cumulative.
  const rowYs: number[] = [];
  let y = canvasMargin + extAboveHeight;
  for (const h of rowHeights) {
    y += Math.floor(h / 2);
    rowYs.push(y);
    y += Math.floor(h / 2) + rowGap;
  }

  // Region centre-X, cumulative.
  const regionXs = new Map<string, number>();
  let x = canvasMargin;
  for (const rid of regionIds) {
    const w = regionWidths.get(rid)!;
    x += Math.floor(w / 2);
    regionXs.set(rid, x);
    x += Math.floor(w / 2) + regionGap;
  }
  const canvasRight = x - regionGap + canvasMargin;

  // Place each component at its grid cell.
  for (const rid of regionIds) {
    const rx = regionXs.get(rid)!;
    const rows = regionLayout[rid];
    for (let i = 0; i < rows.length; i++) {
      const row = rows[i];
      const ry = rowYs[i];
      const totalW = row.reduce((acc, cid) => acc + (sizes.get(cid)?.w ?? 0), 0)
        + (row.length - 1) * cellGap;
      let cursor = rx - Math.floor(totalW / 2);
      for (const cid of row) {
        const cw = sizes.get(cid)?.w ?? 0;
        getComponent(diagram, cid).pos = [cursor + Math.floor(cw / 2), ry];
        cursor += cw + cellGap;
      }
    }
  }

  // Place external (above a target).
  for (const [eid, spec] of Object.entries(ext)) {
    if (spec.above) {
      const target = getComponent(diagram, spec.above);
      const gap = spec.gap ?? 60;
      const targetTop = target.pos[1] - Math.floor((sizes.get(spec.above)?.h ?? 0) / 2);
      getComponent(diagram, eid).pos = [
        target.pos[0],
        targetTop - gap - Math.floor((sizes.get(eid)?.h ?? 0) / 2),
      ];
    }
  }

  // Region bounds — hug the cells with region padding.
  for (const rid of regionIds) {
    const rx = regionXs.get(rid)!;
    const rw = regionWidths.get(rid)!;
    const rows = regionLayout[rid];
    const n = rows.length;
    const top = rowYs[0] - Math.floor(rowHeights[0] / 2) - regionPaddingY;
    const bot = rowYs[n - 1] + Math.floor(rowHeights[n - 1] / 2) + regionPaddingY;
    for (const r of diagram.regions) {
      if (r.id === rid) {
        r.bounds = [rx - Math.floor(rw / 2), top, rw, bot - top];
        break;
      }
    }
  }

  // Auto-route edges that asked for it.
  routeEdges(diagram, rowYs);

  // Final canvas — fit content + margin.
  diagram.width = canvasRight;
  diagram.height = Math.max(
    ...diagram.components.map((c) => c.pos[1] + Math.floor((sizes.get(c.id)?.h ?? 0) / 2)),
  ) + canvasMargin;
}

// ── Anonymous layout-tree (Figma-style auto-layout) ────────────────────
export function applyLayoutTree(
  byId: Map<string, Component>,
  tree: LayoutNode,
  opts: { gap?: number; origin?: Point } = {},
): Point {
  const gap = opts.gap ?? 40;
  const origin = opts.origin ?? [0, 0];

  const lookup = (cid: string): Component => {
    const c = byId.get(cid);
    if (!c) throw new Error(`layout tree references unknown component ${JSON.stringify(cid)}`);
    return c;
  };

  const cell = (c: Component): Point => {
    const [iw, ih] = iconDims(c.shape);
    const lh = (c.name || c.subtitle) ? (LABEL_HEIGHT[c.shape] ?? 0) : 0;
    return [iw, ih + lh];
  };

  const padOf = (node: LayoutNode): Point =>
    (node.t === "group" && node.padding) ? node.padding : [0, 0];

  const effectiveGap = (dir: "horizontal" | "vertical", sizes: Point[]): number => {
    const cross = dir === "horizontal"
      ? Math.max(...sizes.map((s) => s[1]))
      : Math.max(...sizes.map((s) => s[0]));
    return Math.max(gap, Math.floor(cross / 4));
  };

  const measure = (node: LayoutNode): Point => {
    if (node.t === "id") return cell(lookup(node.id));
    const [px, py] = padOf(node);
    const sizes = node.children.map(measure);
    const eg = effectiveGap(node.dir, sizes);
    if (node.dir === "horizontal") {
      return [
        sizes.reduce((a, s) => a + s[0], 0) + (sizes.length - 1) * eg + 2 * px,
        Math.max(...sizes.map((s) => s[1])) + 2 * py,
      ];
    }
    return [
      Math.max(...sizes.map((s) => s[0])) + 2 * px,
      sizes.reduce((a, s) => a + s[1], 0) + (sizes.length - 1) * eg + 2 * py,
    ];
  };

  const place = (node: LayoutNode, cx: number, cy: number): void => {
    if (node.t === "id") {
      const c = lookup(node.id);
      const lh = (c.name || c.subtitle) ? (LABEL_HEIGHT[c.shape] ?? 0) : 0;
      c.pos = [cx, cy - Math.floor(lh / 2)];   // cell center → icon center
      return;
    }
    const sizes = node.children.map(measure);
    const eg = effectiveGap(node.dir, sizes);
    const [w, h] = measure(node);
    const [px, py] = padOf(node);
    const innerW = w - 2 * px;
    const innerH = h - 2 * py;
    if (node.dir === "horizontal") {
      let cursor = cx - Math.floor(innerW / 2);
      node.children.forEach((ch, i) => {
        const cw = sizes[i][0];
        place(ch, cursor + Math.floor(cw / 2), cy);
        cursor += cw + eg;
      });
    } else {
      let cursor = cy - Math.floor(innerH / 2);
      node.children.forEach((ch, i) => {
        const chH = sizes[i][1];
        place(ch, cx, cursor + Math.floor(chH / 2));
        cursor += chH + eg;
      });
    }
  };

  const [w, h] = measure(tree);
  place(tree, origin[0] + Math.floor(w / 2), origin[1] + Math.floor(h / 2));
  return [w, h];
}

// ── Crossing minimisation (barycenter heuristic) ────────────────────────
function collectLeavesOrdered(node: LayoutNode): string[] {
  if (node.t === "id") return [node.id];
  const out: string[] = [];
  for (const ch of node.children) out.push(...collectLeavesOrdered(ch));
  return out;
}

export function minimizeCrossings(
  tree: LayoutNode, edges: Array<[string, string]>, maxSweeps = 24,
): void {
  if (tree.t !== "group") return;

  const leafAdj = new Map<string, Set<string>>();
  const add = (a: string, b: string): void => {
    if (!leafAdj.has(a)) leafAdj.set(a, new Set());
    leafAdj.get(a)!.add(b);
  };
  for (const [s, d] of edges) { add(s, d); add(d, s); }

  for (let i = 0; i < maxSweeps; i++) {
    if (!sweepNode(tree, leafAdj)) break;
  }
}

function sweepNode(node: LayoutNode, leafAdj: Map<string, Set<string>>): boolean {
  if (node.t !== "group") return false;
  let changed = false;
  for (const ch of node.children) {
    if (sweepNode(ch, leafAdj)) changed = true;
  }
  const sibs = node.children;
  const n = sibs.length;
  if (n < 2) return changed;
  for (let i = 0; i < n - 1; i++) {
    if (barycenterReorder(sibs, i, i + 1, leafAdj)) changed = true;
  }
  for (let i = n - 1; i > 0; i--) {
    if (barycenterReorder(sibs, i, i - 1, leafAdj)) changed = true;
  }
  return changed;
}

function barycenterReorder(
  sibs: LayoutNode[], fixedIdx: number, freeIdx: number,
  adj: Map<string, Set<string>>,
): boolean {
  const free = sibs[freeIdx];
  if (free.t !== "group") return false;

  const fixedLeaves = collectLeavesOrdered(sibs[fixedIdx]);
  const fixedPos = new Map<string, number>(fixedLeaves.map((leaf, idx) => [leaf, idx]));

  const keyed = free.children.map((ch, chIdx) => {
    const positions: number[] = [];
    for (const leaf of collectLeavesOrdered(ch)) {
      for (const adjLeaf of adj.get(leaf) ?? []) {
        if (fixedPos.has(adjLeaf)) positions.push(fixedPos.get(adjLeaf)!);
      }
    }
    const bary = positions.length
      ? positions.reduce((a, b) => a + b, 0) / positions.length
      : chIdx;
    return { bary, chIdx, ch };
  });

  keyed.sort((a, b) => (a.bary - b.bary) || (a.chIdx - b.chIdx));   // stable on ties
  const newChildren = keyed.map((t) => t.ch);
  if (newChildren.length === free.children.length
    && newChildren.every((c, i) => c === free.children[i])) {
    return false;
  }
  free.children = newChildren;
  return true;
}

// ── Edge routing helpers ───────────────────────────────────────────────
function routeEdges(diagram: Diagram, rowYs: number[]): void {
  for (const e of diagram.edges) {
    if (e.via.length) continue;            // respect explicit waypoints
    if (e.route === "over") e.via = routeOver(diagram, e, rowYs);
    else if (e.route === "under") e.via = routeUnder(diagram, e, rowYs);
    // "auto" → no via; the renderer's auto-elbow handles it.
  }
}

function nearestRow(rowYs: number[], y: number): number {
  let best = 0;
  let bestD = Infinity;
  for (let i = 0; i < rowYs.length; i++) {
    const d = Math.abs(rowYs[i] - y);
    if (d < bestD) { bestD = d; best = i; }
  }
  return best;
}

function routeOver(diagram: Diagram, e: Edge, rowYs: number[]): Point[] {
  const src = getComponent(diagram, e.src);
  const dst = getComponent(diagram, e.dst);
  const [sa, da] = resolveAnchors(e, src, dst);
  const sp = anchor(src, sa);
  const dp = anchor(dst, da);

  const srcRow = nearestRow(rowYs, src.pos[1]);
  if (srcRow === 0) {
    const tops = diagram.components
      .filter((c) => c.shape !== "annotation")
      .map((c) => c.pos[1] - componentHalf(c)[1]);
    const topOfContent = tops.length ? Math.min(...tops) : 0;
    const overY = Math.max(14, topOfContent - 30);
    return [[sp[0], overY], [dp[0], overY]];
  }

  const TOL = 5;
  const prevRowY = rowYs[srcRow - 1];
  const currRowY = rowYs[srcRow];
  const prevCells = diagram.components.filter((c) => Math.abs(c.pos[1] - prevRowY) < TOL);
  const currCells = diagram.components.filter((c) => Math.abs(c.pos[1] - currRowY) < TOL);
  const prevBottom = prevCells.length
    ? Math.max(...prevCells.map((c) => c.pos[1] + componentHalf(c)[1] + (LABEL_HEIGHT[c.shape] ?? 0)))
    : prevRowY;
  const currTop = currCells.length
    ? Math.min(...currCells.map((c) => c.pos[1] - componentHalf(c)[1]))
    : currRowY;
  const overY = Math.floor((prevBottom + currTop) / 2);
  return [[sp[0], overY], [dp[0], overY]];
}

function routeUnder(diagram: Diagram, e: Edge, rowYs: number[]): Point[] {
  const src = getComponent(diagram, e.src);
  const dst = getComponent(diagram, e.dst);
  const [sa, da] = resolveAnchors(e, src, dst);
  const sp = anchor(src, sa);
  const dp = anchor(dst, da);

  const srcRow = nearestRow(rowYs, src.pos[1]);
  if (srcRow + 1 >= rowYs.length) {
    const yy = Math.max(sp[1], dp[1]) + 36;
    return [[sp[0], yy], [dp[0], yy]];
  }

  const TOL = 5;
  const srcRowY = rowYs[srcRow];
  const nextRowY = rowYs[srcRow + 1];
  const srcCells = diagram.components.filter((c) => Math.abs(c.pos[1] - srcRowY) < TOL);
  const nextCells = diagram.components.filter((c) => Math.abs(c.pos[1] - nextRowY) < TOL);
  const rowBottom = srcCells.length
    ? Math.max(...srcCells.map((c) => c.pos[1] + componentHalf(c)[1] + (LABEL_HEIGHT[c.shape] ?? 0)))
    : srcRowY;
  const nextTop = nextCells.length
    ? Math.min(...nextCells.map((c) => c.pos[1] - componentHalf(c)[1]))
    : nextRowY;
  const underY = Math.floor((rowBottom + nextTop) / 2);
  return [[sp[0], underY], [dp[0], underY]];
}
